archive/zip_archiver.go

```diff
@@ -84,7 +107,38 @@ func checkMatch(fileName string, excludes []string) (value bool) {
 	return false
 }
 
-func (a *ZipArchiver) ArchiveDir(indirname string, excludes []string) error {
+// The basic file header is very simple. The UncompressedSize logic is not a real-world use case
+// in this context, but "640K ought to be enough for anybody".
+//
+// For reference, see golang/src/archive/zip/struct.go.
+func prepareEmptyHeader(content []byte, infilename string) *zip.FileHeader {
+	fh := &zip.FileHeader{
+		Name:               filepath.ToSlash(infilename),
+		UncompressedSize64: uint64(len(content)),
+	}
+
+	if fh.UncompressedSize64 > uint32max {
+		fh.UncompressedSize = uint32max
+	} else {
+		fh.UncompressedSize = uint32(fh.UncompressedSize64)
+	}
+
+	return fh
+}
+
+// Normalize the fields:
+//
+// - no compression, so the compressed stream is essentially a copy;
+// - fixed date;
+// - fixed file permissions.
+//
+func normalizeCompressingFile(fh *zip.FileHeader) {
+	fh.Method = zip.Store
```

Comment:
This causes a very sneaky compatibility issue with Java's standard java.util.zip, resulting in:
```
java.util.zip.ZipException: only DEFLATED entries can have EXT descriptor
        at java.base/java.util.zip.ZipInputStream.readLOC(ZipInputStream.java:311)
        at java.base/java.util.zip.ZipInputStream.getNextEntry(ZipInputStream.java:123)
        at com.meh.test.UnzipFiles.unzip(UnzipFiles.java:30)
        at com.meh.test.UnzipFiles.main(UnzipFiles.java:17)
```

The quick summary seems to be: Go's zip uses streaming, meaning it first writes the file entry body, and then writes a so-called "ext" data descriptor with size information. Then, java.util.zip can extract these if they use DEFLATE, since DEFLATE encodes data in such a way that you can know when you have all the data. STORE, on the other hand, has no such encoded info, and Java fails to handle it.

Details:
* https://bugs.openjdk.java.net/browse/JDK-8143613
* https://github.com/30x/zipper
* https://groups.google.com/g/golang-nuts/c/0iae5Ng-I-0
* https://stackoverflow.com/questions/47208272/android-zipinputstream-only-deflated-entries-can-have-ext-descriptor

Impact case: The _Amazon ECS (Blue/Green)_ deploy provider of AWS CodePipeline raises an exception if a normalized zip file is fed to it:

> Exception while trying to read the task definition artifact file from: \<source stage\>

We use archive_file to produce the zip, containing appspec.yaml and taskdef.json files for CodeDeploy and ECS, respectively.

AWS services seem to be heavily Java based, so while we don't see the full Java exception output, it seems likely that this is the problem.